### PR TITLE
Fix for issue 88 that fragmenter settings in pipeline reflect back to fragmentation settings in main view

### PR DESCRIPTION
**Pull request by JonasSchaub**
Now made sure that fragmenter instances are always copied when used in the pipeline settings view controller. 
Also some renaming to make things more clear and also added a note about pipeline fragmenter settings to the tutorial.

### CODE REVIEW

**Comment by JonasSchaub** [comment]
@FelixBaensch, very sorry to bother you, but could you maybe test and review this. Please forgive me for the refactoring that makes the "actual" changes harder to see.

File: src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java [hunk 1]
```diff
@@ -205,46 +215,48 @@ private void addListenerAndBindings() {
...
-                        this.fragmenterList.set(tmpIndex, Arrays.stream(this.fragmenters)
-                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterString)).findFirst().orElse(null));
+                    if (this.selectedPipelineFragmentersList.size() > tmpIndex) {
+                        this.selectedPipelineFragmentersList.set(tmpIndex, Arrays.stream(this.availableFragmentersMasterArray)
+                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterDisplayName)).findFirst().get().copy());
                     } else {
-                        this.fragmenterList.add(Arrays.stream(this.fragmenters)
-                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterString)).findFirst().orElse(null));
+                        this.selectedPipelineFragmentersList.add(Arrays.stream(this.availableFragmentersMasterArray)
+                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterDisplayName)).findFirst().get().copy());
```
Comment (JonasSchaub) [inline src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java:254]: @FelixBaensch might I ask what is the reason behind this stream filtering here? We already identified the fragmenter to use via the for loop and the "if name equals fragmenter.name" condition, right?

File: src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java [hunk 2]
```diff
@@ -135,18 +136,26 @@ public PipelineSettingsViewController(Stage aMainStage,
...
         this.fragmentationService = aFragmentationService;
-        this.fragmenters = this.fragmentationService.getFragmenters();
-        this.pipelineName = new SimpleStringProperty();
-        this.fragmenterList = new LinkedList<>();
-        // copies pipeline fragmenters from fragmentation service to the internal fragmenter list of this class as initialisation
+        //developers note: this is basically a proxy to make absolutely sure nothing is changed in the fragmentation service
+        this.availableFragmentersMasterArray = new IMoleculeFragmenter[this.fragmentationService.getFragmenters().length];
+        for (int i = 0; i < this.fragmentationService.getFragmenters().length; i++) {
+            this.availableFragmentersMasterArray[i] = this.fragmentationService.getFragmenters()[i].copy();
+        }
```
Comment (JonasSchaub) [inline src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java:145]: Here is one "actual" change.

File: src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java (hunk 1 above)
Comment (JonasSchaub) [reply src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java:254]: And there is also another "actual" change here, the added copy() calls.

File: src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java [hunk 3]
```diff
@@ -253,14 +265,14 @@ private void addNewChoiceRow(String aFragmenterDisplayName, boolean anUpdateFrag
             //does not trigger the action defined above
             tmpComboBox.getSelectionModel().select(aFragmenterDisplayName);
             if (anUpdateFragmentersList) {
-                for (IMoleculeFragmenter tmpFragmenter : this.fragmenters) {
+                for (IMoleculeFragmenter tmpFragmenter : this.availableFragmentersMasterArray) {
                     if (aFragmenterDisplayName.equals(tmpFragmenter.getFragmentationAlgorithmDisplayName())) {
-                        this.fragmenterList.add(Arrays.stream(this.fragmenters).filter(x ->
-                                x.getFragmentationAlgorithmDisplayName().equals(aFragmenterDisplayName)).findFirst().orElse(null));
+                        this.selectedPipelineFragmentersList.add(Arrays.stream(this.availableFragmentersMasterArray).filter(x ->
+                                x.getFragmentationAlgorithmDisplayName().equals(aFragmenterDisplayName)).findFirst().get().copy());
```
Comment (JonasSchaub) [inline src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java:271]: Also copy() added here

**Comment by FelixBaensch** [comment]
> @FelixBaensch, very sorry to bother you, but could you maybe test and review this? Please forgive me for the refactoring that makes the "actual" changes harder to see.

I will leave these extraordinarily long variable names uncommented....
But I will review it in the next few days

File: src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java (hunk 1 above)
Comment (FelixBaensch) [reply src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java:254]: I have double checked this, and If I see it correctly this is your code :)

Comment (FelixBaensch) [reply src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java:254]: Irc, this is legacy code, as we only allowed each fragmenter once per pipeline

**Review by FelixBaensch — COMMENTED** [review]
See questions

**Comment by JonasSchaub** [comment]
@FelixBaensch I have restructured the code and removed everything that I regarded as legacy code. Could you maybe test whether I broke sth and whether the original issue is fixed?

File: src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java (hunk 2 above)
Comment (FelixBaensch) [reply src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java]: ?

**Comment by JonasSchaub** [comment]
@FelixBaensch I have made adjustments according to your requested changes. Do you approve?

**Review by FelixBaensch — APPROVED** [review]

**Comment by FelixBaensch** [comment]
Merge at will

**Pull request merged**